File: src/main/java/raft/war/binary/parser/w3g/parser/replay/ChatRecord.java

package raft.war.binary.parser.w3g.parser.replay;

import raft.war.binary.parser.w3g.parser.packed.IRecord;
import raft.war.binary.parser.w3g.parser.utils.ByteBufferUtil;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.Objects;

public class ChatRecord implements IRecord {

    public static final int TYPE = 0x20;

    private byte playerId;

    private Integer chatMode;

    private byte flags;

    private String message;

    @Override
    public int getRecordId() {
        return TYPE;
    }

    @Override
    public void parse(ByteBuffer inBuffer) {
        playerId = inBuffer.get();
        inBuffer.getShort(); // length
        flags = inBuffer.get();
        chatMode = (flags != 0x10) ? inBuffer.getInt() : null;
        message = ByteBufferUtil.readUtf8CString(inBuffer);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (o == null || getClass() != o.getClass())
            return false;
        ChatRecord that = (ChatRecord) o;
        return playerId == that.playerId && flags == that.flags && Objects.equals(chatMode, that.chatMode) && Objects.equals(message, that.message);
    }

    @Override
    public int hashCode() {
        return Objects.hash(playerId, chatMode, flags, message);
    }

    public byte getPlayerId() {
        return playerId;
    }

    public ChatRecord setPlayerId(byte playerId) {
        this.playerId = playerId;
        return this;
    }

    public Integer getChatMode() {
        return chatMode;
    }

    public ChatRecord setChatMode(Integer chatMode) {
        this.chatMode = chatMode;
        return this;
    }

    public byte getFlags() {
        return flags;
    }

    public ChatRecord setFlags(byte flags) {
        this.flags = flags;
        return this;
    }

    public String getMessage() {
        return message;
    }

    public ChatRecord setMessage(String message) {
        this.message = message;
        return this;
    }
}

File: src/main/kotlin/raft/war/binary/parser/w3g/record/ChatRecord.kt
package raft.war.binary.parser.w3g.record

import raft.war.binary.parser.w3g.parser.utils.ByteBufferUtil
import java.nio.ByteBuffer
import java.util.*

class ChatRecord : RecordBase {
    override var timestamp: Long = 0

    var playerId: Byte = 0
    var chatMode: Int? = null
    var flags: Byte = 0
    var message: String? = null

    override fun parse(inBuffer: ByteBuffer) {
        playerId = inBuffer.get()
        inBuffer.getShort() // length
        flags = inBuffer.get()
        chatMode = if ((flags.toInt() != 0x10)) inBuffer.getInt() else null
        message = ByteBufferUtil.readUtf8CString(inBuffer)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as ChatRecord
        return playerId == that.playerId && flags == that.flags && chatMode == that.chatMode && message == that.message
    }

    override fun hashCode(): Int = Objects.hash(playerId, chatMode, flags, message)

    companion object {
        const val ID: Int = 0x20
    }

    override fun toString(): String = "Time: $timestamp,\tChatRecord" +
            "\tplayerId: $playerId\n" +
            "\tchatMode: $chatMode\n" +
            "\tflags: $flags\n" +
            "\tmessage: ${message?.replace("\r", "\n")}\n"
}
